Enable 2FA when publishing a new package

SamVerschueren opened the pull request:
This PR fixes #288 by adding a new step only if the package is new and the package is not private.

I also moved npm specific parts to an `npm` directory. We might want to do the same for `git`? I can revert it but to me it looks cleaner this way.

I also noticed that almost every time, the OTP of the publish package will be the same as for the 2FA. So maybe we should just try to re-use the first OTP, if that fails, it will still just ask for a new OTP. But in 99% of the use cases, this would only require to enter the OTP once. What do you think?

SamVerschueren commented on source/cli.js:
```diff
@@ -79,6 +79,7 @@ process.on('SIGINT', () => {
 		{
 			...cli.flags,
 			confirm: true,
+			exists: !isAvailable,
```
I need this, and it is returned from the `ui()` step, so makes sense to add it here as well.

SamVerschueren commented on source/index.js:
```diff
@@ -15,7 +15,8 @@ const pkgDir = require('pkg-dir');
 const prerequisiteTasks = require('./prerequisite-tasks');
 const gitTasks = require('./git-tasks');
-const publishTaskHelper = require('./publish-task-helper');
+const publish = require('./npm/publish');
```
I could merge these two into

```js
const {publish, enable2fa} = require('./npm');
```

I don't have a strong opinion, this might be a bit cleaner.

SamVerschueren commented on source/npm/util.js:
```diff
@@ -0,0 +1,29 @@
+const {throwError} = require('rxjs');
+const {catchError} = require('rxjs/operators');
+
+const handleNpmError = (error, task, message, executor) => {
```
I extracted this out of the `publish-task-helper` as it's necessary for every npm command which uses OTP.

SamVerschueren commented on source/prerequisite-tasks.js:
```diff
@@ -15,6 +16,20 @@ module.exports = (input, pkg, options) => {
+				const versions = JSON.parse(await execa.stdout('npm', ['version', '--json']));
+
+				if (version.satisfies(versions.npm, '<6.5.0')) {
+					throw new Error(`npm@${versions.npm} does not support enabling two-factor authentication on new repositories`);
```
Not sure about the error messages. Should we help the user and tell which version he should install? Currently, only `npm@6.5.0` works. `npm@6.8.0` will probably work as well, but it's not released yet.

sindresorhus replied:
> `npm@6.8.0` will probably work as well, but it's not released yet.

It is: https://github.com/npm/cli/releases/tag/v6.8.0

I think we should just drop this error message and require npm@6.8.0 for everyone.

SamVerschueren replied:
Done, good idea. When I pushed this yesterday, v6.8 wasn't released yet so I had to add this because otherwise it was untestable :).

sindresorhus commented:
> So maybe we should just try to re-use the first OTP, if that fails, it will still just ask for a new OTP. But in 99% of the use cases, this would only require to enter the OTP once. What do you think?

👍

SamVerschueren commented:
Because of the npm version check, the tests fail on Travis. Should we do a `npm i -g npm@6.5.0` in the `travis.yml` file as pretest script?

sindresorhus commented:
> Should we do a `npm i -g npm@6.5.0` in the `travis.yml` file as pretest script?

```
npm install --global npm@6.8.0
```

SamVerschueren commented on source/prerequisite-tasks.js:
```diff
@@ -15,6 +16,16 @@ module.exports = (input, pkg, options) => {
+				const versions = JSON.parse(await execa.stdout('npm', ['version', '--json']));
+
+				if (version.satisfies(versions.npm, '<6.8.0')) {
+					throw new Error('Please upgrade to npm@6.8.0 or newer');
```
Not sure about the error message here...

SamVerschueren commented:
- Dropped the npm@6.5.0 check
- Added re-use of the OTP provided in the `npm publish` step
- Made sure npm prerelease tags are correctly handled in the npm version check

sindresorhus commented:
Tested this when publishing a new package now and it worked perfectly 👌🎉

sindresorhus commented:
@SamVerschueren You need to submit the PR URL to IssueHunt to claim the bounty.

itaisteinherz commented:
Why are `source/npm/util.js` and `source/npm-util.js` separate files? I can't see a reason why the latter shouldn't be merged into the former.

sindresorhus commented:
@itaisteinherz I agree, it's weird to have it like that. Let's rename `npm-util.js` to `handle-npm-error.js` and move `npm-util.js` into the `npm` directory.

mehulkar commented:
Hi all, I'm using `np` against an internal, private NPM registry that does not support 2FA. How do I disable this feature? I just updated to `np` 5.x and am unable to publish packages now as it fails the 2FA check

itaisteinherz commented:
@mehulkar See #398 (and #424, #427). If you need this ASAP, feel free to open a PR that addresses https://github.com/sindresorhus/np/issues/398#issuecomment-498765559 and we'll gladly review it and try to get it landed.
